primitives/trie/src/cache/mod.rs

```diff
@@ -0,0 +1,687 @@
+	sync::Arc,
+};
+use trie_db::{node::NodeOwned, CachedValue};
+
+mod shared_cache;
+
+pub use shared_cache::SharedTrieCache;
+
+use self::shared_cache::ValueCacheKeyHash;
+
+const LOG_TARGET: &str = "trie-cache";
+
+/// The size of the cache.
+#[derive(Debug, Clone, Copy)]
+pub enum CacheSize {
+	/// Do not limit the cache size.
+	Unlimited,
+	/// Let the cache in maximum use the given amount of bytes.
+	Maximum(usize),
+}
+
+impl CacheSize {
+	/// Returns `true` if the `current_size` exceeds the allowed size.
+	fn exceeds(&self, current_size: usize) -> bool {
+		match self {
+			Self::Unlimited => false,
+			Self::Maximum(max) => *max < current_size,
+		}
+	}
+}
+
+/// The local trie cache.
+///
+/// This cache should be used per state instance created by the backend. One state instance is
+/// referring to the state of one block. It will cache all the accesses that are done to the state
+/// which could not be fullfilled by the [`SharedTrieCache`]. These locally cached items are merged
+/// back to the shared trie cache when this instance is dropped.
+///
+/// When using [`Self::as_trie_db_cache`] or [`Self::as_trie_db_mut_cache`], it will lock Mutexes.
+/// So, it is important that these methods are not called multiple times, because they otherwise
+/// deadlock.
+pub struct LocalTrieCache<H: Hasher> {
+	/// The shared trie cache that created this instance.
+	shared: SharedTrieCache<H>,
+	/// The local cache for the trie nodes.
+	node_cache: Mutex<HashMap<H::Out, NodeOwned<H::Out>>>,
+	/// Keeps track of all the trie nodes accessed in the shared cache.
+	///
+	/// This will be used to ensure that these nodes are brought to the front of the lru when this
+	/// local instance is merged back to the shared cache.
+	shared_node_cache_access: Mutex<HashSet<H::Out>>,
+	/// The local cache for the values.
+	value_cache: Mutex<
+		HashMap<
+			ValueCacheKey<'static, H::Out>,
+			CachedValue<H::Out>,
+			BuildNoHashHasher<ValueCacheKey<'static, H::Out>>,
+		>,
+	>,
+	/// Keeps track of all values accessed in the shared cache.
+	///
+	/// This will be used to ensure that these nodes are brought to the front of the lru when this
+	/// local instance is merged back to the shared cache. This can actually lead to collision when
+	/// two [`ValueCacheKey`]s with different storage roots and keys map to the same hash. However,
+	/// as we only use this set to update the lru position it is fine, even if we bring the wrong
+	/// value to the top. The important part is that we always get the correct value from the value
+	/// cache for a given key.
+	shared_value_cache_access:
+		Mutex<HashSet<ValueCacheKeyHash, BuildNoHashHasher<ValueCacheKeyHash>>>,
+}
+
+impl<H: Hasher> LocalTrieCache<H> {
+	/// Return self as a [`TrieDB`](trie_db::TrieDB) compatible cache.
+	///
+	/// The given `storage_root` needs to be the storage root of the trie this cache is used for.
+	pub fn as_trie_db_cache<'a>(&'a self, storage_root: H::Out) -> TrieCache<'a, H> {
```

Comment:
Can this take `&mut self` so that there's need for mutexes?

Reply:
A bit out of context, but I remember writing some branches where I did switch all state_machine backends and externalities to `&mut self` so we can register access without inner mutablility, I remember it being a lot smoother than expected.

Reply:
I would really like to move on with this :grimacing: I mean yes, removing the `Mutex` would be nice and I can create a followup issue. However, given that we don't have the `Mutex` shared. I don't think that we see any performance problems because of this at the moment.